File: core/fiscal-scope.ts
import { createHash } from "node:crypto";
import { isValidFiscalRuc } from "./ruc.js";
import type { ReviewAcknowledgement } from "./types.js";

export const FISCAL_SCOPE_VERSION = { V1: "v1" } as const;
export type FiscalScopeVersion =
	(typeof FISCAL_SCOPE_VERSION)[keyof typeof FISCAL_SCOPE_VERSION];
export const AUTHORITY_LEVEL = {
	ASK: "ASK",
	ANALYZE: "ANALYZE",
	PREPARE: "PREPARE",
	EXECUTE: "EXECUTE",
} as const;
export type AuthorityLevel =
	(typeof AUTHORITY_LEVEL)[keyof typeof AUTHORITY_LEVEL];
export const FISCAL_SCOPE_ERROR = {
	REQUIRED: "SCOPE_BINDING_REQUIRED",
	INVALID: "SCOPE_BINDING_INVALID",
	MISMATCH: "SCOPE_MISMATCH",
	UNSUPPORTED_VERSION: "UNSUPPORTED_SCOPE_VERSION",
} as const;
export type FiscalScopeErrorCode =
	(typeof FISCAL_SCOPE_ERROR)[keyof typeof FISCAL_SCOPE_ERROR];

export class FiscalScopeError extends Error {
	constructor(
		readonly code: FiscalScopeErrorCode,
		message: string,
	) {
		super(`${code}: ${message}`);
	}
}

export interface FiscalScopeBinding {
	version: FiscalScopeVersion;
	tenant: string;
	organization: string;
	company: string;
	fiscalPeriod: string;
	ledgerBook: string;
	operationType: string;
	sourceSnapshot: string;
	policyVersion: string;
	actor: string;
	authorityLevel: AuthorityLevel;
}

const KEYS = [
	"version",
	"tenant",
	"organization",
	"company",
	"fiscalPeriod",
	"ledgerBook",
	"operationType",
	"sourceSnapshot",
	"policyVersion",
	"actor",
	"authorityLevel",
] as const;
const OPERATIONS: Readonly<Record<string, AuthorityLevel>> = {
	"memory.save": AUTHORITY_LEVEL.PREPARE,
	"memory.supersede": AUTHORITY_LEVEL.PREPARE,
	"evidence.store": AUTHORITY_LEVEL.PREPARE,
	"evidence.link": AUTHORITY_LEVEL.PREPARE,
	"rule.link": AUTHORITY_LEVEL.PREPARE,
	"close.create": AUTHORITY_LEVEL.PREPARE,
	"context.read": AUTHORITY_LEVEL.ASK,
	"review.queue": AUTHORITY_LEVEL.ASK,
	"review.detail": AUTHORITY_LEVEL.ASK,
	"timeline.read": AUTHORITY_LEVEL.ASK,
	"evidence.get": AUTHORITY_LEVEL.ASK,
	"reconstruct.read": AUTHORITY_LEVEL.ANALYZE,
	"verify.read": AUTHORITY_LEVEL.ANALYZE,
	"memory.approve": AUTHORITY_LEVEL.EXECUTE,
	"close.approve": AUTHORITY_LEVEL.EXECUTE,
	"period.reopen": AUTHORITY_LEVEL.EXECUTE,
	"period.write-check": AUTHORITY_LEVEL.EXECUTE,
};

export function decodeFiscalScopeV1(input: unknown): FiscalScopeBinding {
	if (typeof input !== "object" || input === null || Array.isArray(input))
		fail(FISCAL_SCOPE_ERROR.INVALID, "binding must be one object");
	const record = input as Record<string, unknown>;
	if (!("version" in record))
		fail(FISCAL_SCOPE_ERROR.REQUIRED, "binding version is required");
	if (record.version !== FISCAL_SCOPE_VERSION.V1)
		fail(
			FISCAL_SCOPE_ERROR.UNSUPPORTED_VERSION,
			"unsupported fiscal scope version",
		);
	const keys = Object.keys(record);
	if (
		keys.length !== KEYS.length ||
		keys.some((key, index) => key !== KEYS[index])
	)
		fail(
			FISCAL_SCOPE_ERROR.INVALID,
			"keys must be unique and in canonical order",
		);
	for (const key of KEYS)
		if (typeof record[key] !== "string")
			fail(FISCAL_SCOPE_ERROR.INVALID, `${key} must be a string`);
	// SAFETY: exact keys, string values, and version were checked above; closed vocabularies are checked below.
	const binding = record as unknown as FiscalScopeBinding;
	for (const key of KEYS.slice(1)) {
		const value = binding[key];
		if (
			!value.trim() ||
			value.includes("\0") ||
			!isValidUnicodeScalarString(value)
		)
			fail(FISCAL_SCOPE_ERROR.INVALID, `${key} is invalid`);
	}
	if (!isValidFiscalRuc(binding.company))
		fail(FISCAL_SCOPE_ERROR.INVALID, "company is not a valid RUC");
	if (!/^\d{4}(0[1-9]|1[0-2])$/.test(binding.fiscalPeriod))
		fail(FISCAL_SCOPE_ERROR.INVALID, "fiscalPeriod must be YYYYMM");
	if (!/^[0-9a-f]{64}$/.test(binding.sourceSnapshot))
		fail(FISCAL_SCOPE_ERROR.INVALID, "sourceSnapshot must be lowercase sha256");
	if (OPERATIONS[binding.operationType] !== binding.authorityLevel)
		fail(
			FISCAL_SCOPE_ERROR.INVALID,
			"operationType and authorityLevel do not match the v1 vocabulary",
		);
	return { ...binding };
}

export function parseFiscalScopeV1JSON(text: string): FiscalScopeBinding {
	let cursor = skip(text, 0);
	if (text[cursor++] !== "{")
		fail(FISCAL_SCOPE_ERROR.INVALID, "binding must be one JSON object");
	const values: Record<string, string> = {};
	for (const [index, expected] of KEYS.entries()) {
		cursor = skip(text, cursor);
		if (text[cursor] === "}")
			fail(FISCAL_SCOPE_ERROR.REQUIRED, "complete v1 binding is required");
		const key = readString(text, cursor);
		cursor = skip(text, key.next);
		if (key.value !== expected)
			fail(
				FISCAL_SCOPE_ERROR.INVALID,
				"keys must be unique and in canonical order",
			);
		if (text[cursor++] !== ":")
			fail(FISCAL_SCOPE_ERROR.INVALID, "invalid binding JSON");
		const value = readString(text, skip(text, cursor));
		cursor = skip(text, value.next);
		values[key.value] = value.value;
		if (index === 0 && value.value !== FISCAL_SCOPE_VERSION.V1)
			fail(
				FISCAL_SCOPE_ERROR.UNSUPPORTED_VERSION,
				"unsupported fiscal scope version",
			);
		const last = index === KEYS.length - 1;
		if (text[cursor++] !== (last ? "}" : ","))
			fail(
				last ? FISCAL_SCOPE_ERROR.INVALID : FISCAL_SCOPE_ERROR.REQUIRED,
				"complete canonical binding is required",
			);
	}
	if (skip(text, cursor) !== text.length)
		fail(FISCAL_SCOPE_ERROR.INVALID, "trailing JSON value");
	return decodeFiscalScopeV1(values);
}

export function canonicalFiscalScopeBytes(
	binding: FiscalScopeBinding,
): Uint8Array {
	const checked = decodeFiscalScopeV1(binding);
	let canonical = "drenyra:fiscal-scope:v1\0";
	for (const key of KEYS.slice(1)) {
		const value = checked[key];
		canonical += `${key}=${new TextEncoder().encode(value).length}:${value}\0`;
	}
	return new TextEncoder().encode(canonical);
}

export function fiscalScopeHash(binding: FiscalScopeBinding): string {
	return createHash("sha256")
		.update(canonicalFiscalScopeBytes(binding))
		.digest("hex");
}

export const REVIEW_ACKNOWLEDGEMENT_STATE = {
	OMITTED: "omitted",
	FALSE: "false",
	TRUE: "true",
} as const;
export type ReviewAcknowledgementState =
	(typeof REVIEW_ACKNOWLEDGEMENT_STATE)[keyof typeof REVIEW_ACKNOWLEDGEMENT_STATE];
export function reviewAcknowledgementState(
	value: ReviewAcknowledgement,
): ReviewAcknowledgementState {
	return value.present
		? value.value
			? REVIEW_ACKNOWLEDGEMENT_STATE.TRUE
			: REVIEW_ACKNOWLEDGEMENT_STATE.FALSE
		: REVIEW_ACKNOWLEDGEMENT_STATE.OMITTED;
}

function readString(
	text: string,
	start: number,
): { value: string; next: number } {
	if (text[start] !== '"')
		fail(FISCAL_SCOPE_ERROR.INVALID, "binding keys and values must be strings");
	for (let cursor = start + 1; cursor < text.length; cursor++) {
		if (text[cursor] === "\\") {
			cursor++;
			continue;
		}
		if (text[cursor] === '"') {
			try {
				return {
					value: JSON.parse(text.slice(start, cursor + 1)) as string,
					next: cursor + 1,
				};
			} catch {
				fail(FISCAL_SCOPE_ERROR.INVALID, "invalid JSON string");
			}
		}
	}
	return fail(FISCAL_SCOPE_ERROR.INVALID, "unterminated JSON string");
}
function skip(text: string, start: number): number {
	while (/\s/.test(text[start] ?? "")) start++;
	return start;
}
function isValidUnicodeScalarString(value: string): boolean {
	try {
		encodeURIComponent(value);
		return true;
	} catch {
		return false;
	}
}
function fail(code: FiscalScopeErrorCode, message: string): never {
	throw new FiscalScopeError(code, message);
}

File: internal/core/fiscal_scope.go
package core

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"strconv"
	"strings"
	"unicode/utf8"
)

const (
	ScopeBindingRequired    = "SCOPE_BINDING_REQUIRED"
	ScopeBindingInvalid     = "SCOPE_BINDING_INVALID"
	ScopeMismatch           = "SCOPE_MISMATCH"
	UnsupportedScopeVersion = "UNSUPPORTED_SCOPE_VERSION"
)

type FiscalScopeError struct{ Code, Message string }

func (e *FiscalScopeError) Error() string { return e.Code + ": " + e.Message }

type AuthorityLevel string

const (
	AuthorityLevelAsk     AuthorityLevel = "ASK"
	AuthorityLevelAnalyze AuthorityLevel = "ANALYZE"
	AuthorityLevelPrepare AuthorityLevel = "PREPARE"
	AuthorityLevelExecute AuthorityLevel = "EXECUTE"
)

type FiscalScopeBinding struct {
	Version        string         `json:"version"`
	Tenant         string         `json:"tenant"`
	Organization   string         `json:"organization"`
	Company        string         `json:"company"`
	FiscalPeriod   string         `json:"fiscalPeriod"`
	LedgerBook     string         `json:"ledgerBook"`
	OperationType  string         `json:"operationType"`
	SourceSnapshot string         `json:"sourceSnapshot"`
	PolicyVersion  string         `json:"policyVersion"`
	Actor          string         `json:"actor"`
	AuthorityLevel AuthorityLevel `json:"authorityLevel"`
}

var fiscalScopeKeys = [...]string{"version", "tenant", "organization", "company", "fiscalPeriod", "ledgerBook", "operationType", "sourceSnapshot", "policyVersion", "actor", "authorityLevel"}
var fiscalOperations = map[string]AuthorityLevel{
	"memory.save": AuthorityLevelPrepare, "memory.supersede": AuthorityLevelPrepare,
	"evidence.store": AuthorityLevelPrepare, "evidence.link": AuthorityLevelPrepare,
	"rule.link": AuthorityLevelPrepare, "close.create": AuthorityLevelPrepare,
	"context.read": AuthorityLevelAsk, "review.queue": AuthorityLevelAsk,
	"review.detail": AuthorityLevelAsk, "timeline.read": AuthorityLevelAsk, "evidence.get": AuthorityLevelAsk,
	"reconstruct.read": AuthorityLevelAnalyze, "verify.read": AuthorityLevelAnalyze,
	"memory.approve": AuthorityLevelExecute, "close.approve": AuthorityLevelExecute,
	"period.reopen": AuthorityLevelExecute, "period.write-check": AuthorityLevelExecute,
}

func DecodeFiscalScopeV1JSON(raw []byte) (FiscalScopeBinding, error) {
	if !utf8.Valid(raw) {
		return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "binding must be valid UTF-8")
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	first, err := decoder.Token()
	if err != nil || first != json.Delim('{') {
		return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "binding must be one JSON object")
	}
	values := make([]string, len(fiscalScopeKeys))
	for i, expected := range fiscalScopeKeys {
		if !decoder.More() {
			if i == 1 && values[0] != "v1" {
				return FiscalScopeBinding{}, scopeError(UnsupportedScopeVersion, "unsupported fiscal scope version")
			}
			return FiscalScopeBinding{}, scopeError(ScopeBindingRequired, "complete v1 binding is required")
		}
		key, err := decoder.Token()
		if err != nil || key != expected {
			return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "keys must be unique and in canonical order")
		}
		if err := decoder.Decode(&values[i]); err != nil {
			return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "all binding values must be strings")
		}
		if i == 0 && values[0] != "v1" {
			return FiscalScopeBinding{}, scopeError(UnsupportedScopeVersion, "unsupported fiscal scope version")
		}
	}
	if decoder.More() {
		return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "unknown binding field")
	}
	if end, err := decoder.Token(); err != nil || end != json.Delim('}') {
		return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "invalid binding object")
	}
	if _, err := decoder.Token(); err != io.EOF {
		return FiscalScopeBinding{}, scopeError(ScopeBindingInvalid, "trailing JSON value")
	}
	result := FiscalScopeBinding{
		Version:        values[0],
		Tenant:         values[1],
		Organization:   values[2],
		Company:        values[3],
		FiscalPeriod:   values[4],
		LedgerBook:     values[5],
		OperationType:  values[6],
		SourceSnapshot: values[7],
		PolicyVersion:  values[8],
		Actor:          values[9],
		AuthorityLevel: AuthorityLevel(values[10]),
	}
	if err := ValidateFiscalScopeBinding(result); err != nil {
		return FiscalScopeBinding{}, err
	}
	return result, nil
}

func ValidateFiscalScopeBinding(binding FiscalScopeBinding) error {
	if binding.Version == "" {
		return scopeError(ScopeBindingRequired, "binding version is required")
	}
	if binding.Version != "v1" {
		return scopeError(UnsupportedScopeVersion, "unsupported fiscal scope version")
	}
	for key, value := range map[string]string{
		"tenant": binding.Tenant, "organization": binding.Organization, "company": binding.Company,
		"fiscalPeriod": binding.FiscalPeriod, "ledgerBook": binding.LedgerBook,
		"operationType": binding.OperationType, "sourceSnapshot": binding.SourceSnapshot,
		"policyVersion": binding.PolicyVersion, "actor": binding.Actor, "authorityLevel": string(binding.AuthorityLevel),
	} {
		if !utf8.ValidString(value) || strings.TrimSpace(value) == "" || strings.ContainsRune(value, 0) {
			return scopeError(ScopeBindingInvalid, key+" is invalid")
		}
	}
	if !IsValidFiscalRUC(binding.Company) {
		return scopeError(ScopeBindingInvalid, "company is not a valid RUC")
	}
	if !IsValidPeriod(binding.FiscalPeriod) {
		return scopeError(ScopeBindingInvalid, "fiscalPeriod must be YYYYMM")
	}
	if len(binding.SourceSnapshot) != 64 {
		return scopeError(ScopeBindingInvalid, "sourceSnapshot must be lowercase sha256")
	}
	if _, err := hex.DecodeString(binding.SourceSnapshot); err != nil || strings.ToLower(binding.SourceSnapshot) != binding.SourceSnapshot {
		return scopeError(ScopeBindingInvalid, "sourceSnapshot must be lowercase sha256")
	}
	expected, ok := fiscalOperations[binding.OperationType]
	if !ok || expected != binding.AuthorityLevel {
		return scopeError(ScopeBindingInvalid, "operationType and authorityLevel do not match the v1 vocabulary")
	}
	return nil
}

func CanonicalFiscalScopeBytes(binding FiscalScopeBinding) []byte {
	values := [...]string{binding.Tenant, binding.Organization, binding.Company, binding.FiscalPeriod, binding.LedgerBook, binding.OperationType, binding.SourceSnapshot, binding.PolicyVersion, binding.Actor, string(binding.AuthorityLevel)}
	var out bytes.Buffer
	out.WriteString("drenyra:fiscal-scope:v1\x00")
	for i, value := range values {
		out.WriteString(fiscalScopeKeys[i+1])
		out.WriteByte('=')
		out.WriteString(strconv.Itoa(len([]byte(value))))
		out.WriteByte(':')
		out.WriteString(value)
		out.WriteByte(0)
	}
	return out.Bytes()
}

func FiscalScopeHash(binding FiscalScopeBinding) string {
	sum := sha256.Sum256(CanonicalFiscalScopeBytes(binding))
	return hex.EncodeToString(sum[:])
}

func scopeError(code, message string) error { return &FiscalScopeError{Code: code, Message: message} }
